platform-sdk/swirlds-common/src/main/java/com/swirlds/common/merkle/proof/StateProof.java

```diff
@@ -0,0 +1,198 @@
+    private StateProofNode root;
+
+    /**
+     * A payload is a leaf node in the merkle tree that is being proven. There will be exactly one payload for each
+     * merkle leaf that is being proven.
+     */
+    private List<MerkleLeaf> payloads;
+
+    private byte[] hashBytes;
+
+    /**
+     * Zero arg constructor required by the serialization framework.
+     */
+    public StateProof() {}
+
+    /**
+     * Create a state proof on the given merkle node.
+     *
+     * @param cryptography provides cryptographic primitives
+     * @param merkleRoot   the root of the merkle tree to create a state proof on
+     * @param signatures   signatures on the root hash of the merkle tree
+     * @param payloads     one or more leaf nodes to create a state proof on, may not contain null leaves
+     */
+    public StateProof(
+            @NonNull final Cryptography cryptography,
+            @NonNull final MerkleNode merkleRoot,
+            @NonNull final Map<NodeId, Signature> signatures,
+            @NonNull final List<MerkleLeaf> payloads) {
+
+        Objects.requireNonNull(cryptography);
+        Objects.requireNonNull(merkleRoot);
+        Objects.requireNonNull(signatures);
+        Objects.requireNonNull(payloads);
+
+        this.payloads = validatePayloads(payloads);
+        this.signatures = processSignatures(signatures);
+        this.root = buildStateProofTree(cryptography, merkleRoot, payloads);
+    }
+
+    /**
+     * Cryptographically validate this state proof using the provided threshold.
+     *
+     * @param cryptography provides cryptographic primitives
+     * @param addressBook  the address book to use to validate the state proof
+     * @param threshold    the threshold of signatures required to trust this state proof
+     * @return true if this state proof is valid, otherwise false
+     * @throws IllegalStateException if this method is called before this object has been fully deserialized
+     */
+    public boolean isValid(
+            @NonNull final Cryptography cryptography,
+            @NonNull final AddressBook addressBook,
+            @NonNull final Threshold threshold) {
+
+        return isValid(cryptography, addressBook, threshold, Signature::verifySignature);
+    }
+
+    /**
+     * Cryptographically validate this state proof using the provided threshold.
+     * <p>
+     * This method is package private to permit test code to provide custom signature validation.
+     *
+     * @param cryptography      provides cryptographic primitives
+     * @param addressBook       the address book to use to validate the state proof
+     * @param threshold         the threshold of signatures required to trust this state proof
+     * @param signatureVerifier a function that verifies a signature
+     * @return true if this state proof is valid, otherwise false
+     * @throws IllegalStateException if this method is called before this object has been fully deserialized
+     */
+    public boolean isValid(
```

Comment:
Could be static or did I miss something?

Reply:
This method acts private on the following class member variables:

- hashBytes
- signatures
- root